modules/javafx.graphics/src/main/native-glass/win/RoActivationSupport.cpp

```diff
@@ -0,0 +1,261 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "RoActivationSupport.h"
+#include <comdef.h>
+#include <winstring.h>
+#include <winnt.h>
+#include <jni.h>
+
+namespace
+{
+    bool initialized = false;
+    const char* moduleNotFoundMessage = "WinRT: %s not found\n";
+
+    const char* catStrW(const char* s1, const wchar_t* s2w)
+    {
+        int s1_len = int(strlen(s1));
+        int s2_len = WideCharToMultiByte(CP_ACP, 0, s2w, -1, NULL, 0, NULL, FALSE);
+        char* res = new char[s1_len + s2_len];
+        WideCharToMultiByte(CP_ACP, 0, s2w, -1, res + s1_len, s2_len, NULL, FALSE);
+        memcpy_s(res, s1_len + s2_len, s1, s1_len);
+        return res;
+    }
+
+    typedef HRESULT WINAPI FnRtlGetVersion(PRTL_OSVERSIONINFOW lpVersionInformation);
+    typedef HRESULT WINAPI FnRoInitialize(RO_INIT_TYPE initType);
+    typedef void WINAPI FnRoUninitialize();
+    typedef HRESULT WINAPI FnRoActivateInstance(HSTRING activatableClassId, IInspectable** instance);
+    typedef HRESULT WINAPI FnWindowsCreateString(PCNZWCH sourceString, UINT32 length, HSTRING* string);
+    typedef HRESULT WINAPI FnWindowsDeleteString(HSTRING string);
+
+    HMODULE hLibComBase = NULL;
+    FnRoInitialize* pRoInitialize = NULL;
+    FnRoUninitialize* pRoUninitialize = NULL;
+    FnRoActivateInstance* pRoActivateInstance = NULL;
+    FnWindowsCreateString* pWindowsCreateString = NULL;
+    FnWindowsDeleteString* pWindowsDeleteString = NULL;
+
+    template<class T>
+    bool loadFunction(HMODULE lib, T*& fnptr, const char* name)
+    {
+        fnptr = reinterpret_cast<T*>(GetProcAddress(lib, name));
+        if (fnptr == nullptr) {
+            fprintf(stderr, "GetProcAddress: %s not loaded\n", name);
+            initialized = false;
+            return false;
+        }
+
+        return true;
+    }
+}
+
+void tryInitializeRoActivationSupport()
+{
+    wchar_t path[MAX_PATH];
+    wchar_t file[MAX_PATH];
+
+    if (GetSystemDirectory(path, sizeof(path) / sizeof(wchar_t)) == 0) {
+        return;
+    }
+
+    memcpy_s(file, sizeof(file), path, sizeof(path));
+    wcscat_s(file, MAX_PATH-1, L"\\ntdll.dll");
+    HMODULE hLibNtdll = LoadLibraryW(file);
+    if (!hLibNtdll) {
+        fprintf(stderr, moduleNotFoundMessage, "ntdll.dll");
+        return;
+    }
+
+    FnRtlGetVersion* pRtlGetVersion;
+    if (!loadFunction(hLibNtdll, pRtlGetVersion, "RtlGetVersion")) {
```

Comment:
Is this really the right way to get the Windows version these days? It used to be that everything in NTDLL was off limits. Seems there's an API now for detecting if you're on Windows 8+:

https://docs.microsoft.com/en-us/windows/win32/sysinfo/version-helper-apis